Mark aten ops as canonical

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86215

This is the first batch of canonical aten ops. 87 in total. More to come in the future PRs. 

native_dropout
abs
add.Tensor
add.Scalar
arange.start_step
bitwise_not
bmm
cat
clamp
constant_pad_nd
convolution
convolution_backward
div.Tensor
div.Scalar
embedding_dense_backward
erf
exp
expand
fill.Scalar
grid_sampler_2d
native_group_norm
native_group_norm_backward
native_layer_norm
native_layer_norm_backward
log
_log_softmax
max.dim
amax
mean.dim
min.dim
amin
mm
mul.Tensor
mul.Scalar
native_batch_norm
permute
scalar_tensor
reciprocal
neg
repeat
relu
gelu
rsqrt
sigmoid
slice.Tensor
slice_scatter
_softmax
squeeze.dim
sum.dim_IntList
sqrt
tanh
unsqueeze
var.dim
where.self
clone
sub.Tensor
sub.Scalar
addmm
_to_copy
view
scatter_add
bitwise_and.Tensor
bitwise_or.Tensor
eq.Scalar
ge.Scalar
le.Scalar
gt.Scalar
lt.Scalar
index_select
nonzero
gather
maximum
minimum
pow.Tensor_Scalar
hardtanh
leaky_relu
_adaptive_avg_pool2d
_adaptive_avg_pool2d_backward
avg_pool2d
avg_pool2d_backward
max_pool2d_with_indices
max_pool2d_with_indices_backward
upsample_bilinear2d.vec
upsample_bilinear2d_backward.vec
upsample_nearest2d.vec
upsample_nearest2d_backward.vec
col2im

SherlockNoMad committed:
Mark aten ops as canonical

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -30,3 +30,6 @@
           across different runs of an operator with identical inputs.
+- tag: canonical
+  desc: |
+          canonical aten ops
```
ya gotta give more description here lol

SherlockNoMad replied:
lol. Updated.

SherlockNoMad committed:
Update on "Mark aten ops as canonical"


This is the first batch of canonical aten ops. 87 in total. More to come in the future PRs. 

native_dropout
abs
add.Tensor
add.Scalar
arange.start_step
bitwise_not
bmm
cat
clamp
constant_pad_nd
convolution
convolution_backward
div.Tensor
div.Scalar
embedding_dense_backward
erf
exp
expand
fill.Scalar
grid_sampler_2d
native_group_norm
native_group_norm_backward
native_layer_norm
native_layer_norm_backward
log
_log_softmax
max.dim
amax
mean.dim
min.dim
amin
mm
mul.Tensor
mul.Scalar
native_batch_norm
permute
scalar_tensor
reciprocal
neg
repeat
relu
gelu
rsqrt
sigmoid
slice.Tensor
slice_scatter
_softmax
squeeze.dim
sum.dim_IntList
sqrt
tanh
unsqueeze
var.dim
where.self
clone
sub.Tensor
sub.Scalar
addmm
_to_copy
view
scatter_add
bitwise_and.Tensor
bitwise_or.Tensor
eq.Scalar
ge.Scalar
le.Scalar
gt.Scalar
lt.Scalar
index_select
nonzero
gather
maximum
minimum
pow.Tensor_Scalar
hardtanh
leaky_relu
_adaptive_avg_pool2d
_adaptive_avg_pool2d_backward
avg_pool2d
avg_pool2d_backward
max_pool2d_with_indices
max_pool2d_with_indices_backward
upsample_bilinear2d.vec
upsample_bilinear2d_backward.vec
upsample_nearest2d.vec
upsample_nearest2d_backward.vec
col2im


[ghstack-poisoned]

nkaretnikov commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -30,3 +30,13 @@
+          In contrast to primTorch, canonical aten opset doesn't decompose ops into explicit
+          type promotion and boardcasting ops.
+          Canonical aten ops is also effectively the opset produced by torchdynamo.export(aten_graph=True),
+          and thus can bse used as an opset for export purpose.
```
came across this issue by chance, noticed a typo here: bse

nkaretnikov commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -30,3 +30,13 @@
+  desc: |
+          Canonical aten ops is a subset of aten ops that remains after aten-to-aten decomposition and
+          functionalization pass. Canonical aten ops are fully functional, and adhere to single static
+          assignment (SSA): this implies there will be no `inplace`, or `_out` variants in this opset.
```
nit: extra comma before "or"

nkaretnikov commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -30,3 +30,13 @@
+- tag: canonical
+  desc: |
+          Canonical aten ops is a subset of aten ops that remains after aten-to-aten decomposition and
+          functionalization pass. Canonical aten ops are fully functional, and adhere to single static
```
nit: extra comma before "and"

anjali411 reviewed (commented):
We should add a test to ensure we are correctly tagging all the relevant ops https://github.com/pytorch/pytorch/blob/a45fead623e3f9a11bacbf5d49b252c3e867167d/test/test_ops.py#L1557
Maybe we can verify if all the ops in the list: `torchdynamo.export(aten_graph=True)` are tagged `canonical`?

SherlockNoMad commented:
> We should add a test to ensure we are correctly tagging all the relevant ops
> 
> https://github.com/pytorch/pytorch/blob/a45fead623e3f9a11bacbf5d49b252c3e867167d/test/test_ops.py#L1557
> 
> 
> Maybe we can verify if all the ops in the list: `torchdynamo.export(aten_graph=True)` are tagged `canonical`?

This is a great suggestions and something I can do in the following PRs.
I am thinking about running  torchdynamo.export() on OpInfo test cases, and assert the ops in the resulting graph are all tagged with canononical.

I won't be able to do it in this PR thought, as this is an incomplete list and many more ops will need to be tagged.

SherlockNoMad commented:
> > This is the first batch of canonical aten ops
> 
> Curious what's the reason for separating batches?

No particular reason :) 
I am sure that more ops will need to be added in the short coming future.

SherlockNoMad committed:
Update on "Mark aten ops as canonical"


This is the first batch of canonical aten ops. 87 in total. More to come in the future PRs. 

native_dropout
abs
add.Tensor
add.Scalar
arange.start_step
bitwise_not
bmm
cat
clamp
constant_pad_nd
convolution
convolution_backward
div.Tensor
div.Scalar
embedding_dense_backward
erf
exp
expand
fill.Scalar
grid_sampler_2d
native_group_norm
native_group_norm_backward
native_layer_norm
native_layer_norm_backward
log
_log_softmax
max.dim
amax
mean.dim
min.dim
amin
mm
mul.Tensor
mul.Scalar
native_batch_norm
permute
scalar_tensor
reciprocal
neg
repeat
relu
gelu
rsqrt
sigmoid
slice.Tensor
slice_scatter
_softmax
squeeze.dim
sum.dim_IntList
sqrt
tanh
unsqueeze
var.dim
where.self
clone
sub.Tensor
sub.Scalar
addmm
_to_copy
view
scatter_add
bitwise_and.Tensor
bitwise_or.Tensor
eq.Scalar
ge.Scalar
le.Scalar
gt.Scalar
lt.Scalar
index_select
nonzero
gather
maximum
minimum
pow.Tensor_Scalar
hardtanh
leaky_relu
_adaptive_avg_pool2d
_adaptive_avg_pool2d_backward
avg_pool2d
avg_pool2d_backward
max_pool2d_with_indices
max_pool2d_with_indices_backward
upsample_bilinear2d.vec
upsample_bilinear2d_backward.vec
upsample_nearest2d.vec
upsample_nearest2d_backward.vec
col2im


[ghstack-poisoned]

ezyang commented:
Don't block on me, but I'm wondering if manually labeling these as opposed to automatically deriving the list (and maybe blocklisting ops that we intend to fix) is better. Certainly computing the post functional ops is quite easy; removing composites is also easy.

anjali411 commented:
> > We should add a test to ensure we are correctly tagging all the relevant ops
> > https://github.com/pytorch/pytorch/blob/a45fead623e3f9a11bacbf5d49b252c3e867167d/test/test_ops.py#L1557
> > 
> > Maybe we can verify if all the ops in the list: `torchdynamo.export(aten_graph=True)` are tagged `canonical`?
> 
> This is a great suggestions and something I can do in the following PRs. I am thinking about running torchdynamo.export() on OpInfo test cases, and assert the ops in the resulting graph are all tagged with canononical.
> 
> I won't be able to do it in this PR thought, as this is an incomplete list and many more ops will need to be tagged.

Yeah but it would also just give you the rest of the ops that need to be labeled. It's ok to add the test in a follow up PR if that's more convenient though.

SherlockNoMad commented:
> these as opposed to automatically deriving the list (and maybe blocklisting ops that we intend to fix) is better. Certainly computing

I think it's a good idea to automatically derive the list (maybe generated through dynamo.export with opinfo tests), and cross check with the tagged ops, as suggested by Anjali.
But I do want this opset to be reviewed and guarded by a dev, so that it's relative stable.

SherlockNoMad commented:
@pytorchbot merge -g

ezyang commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -30,3 +30,13 @@
+          assignment (SSA): this implies there will be no `inplace` or `_out` variants in this opset.
+          This opset is designed to serve as the functional IR to interface with compiler backends.
+          In contrast to primTorch, canonical aten opset doesn't decompose ops into explicit
+          type promotion and boardcasting ops.
```
Typo: broadcasting

ezyang commented:
The text description doesn't say what exactly the dev review should do, so it's incomplete ;)

eellison commented:
Any plans on expanding this ? As it stands, it's a bit misleading.
